### ISSUE TITLE
Errors when creating new analyzer project in VS2017 RC

### ISSUE THREAD

**Issue by bkoelman** [dotnet/roslyn#15978]
**Version Used**: 
Visual Studio 2017 RC, v15.0.26014.0 with NuGet version of Microsoft.CodeAnalysis 2.0.0-rc2

**Steps to Reproduce**:
1. File, New Project, Extensibility, Analyzer with Code Fix
2. While creating the project, an error dialog is shown:
```A problem was encountered creating the sub project 'SomeAnalyzer1.Vsix'.  The parameter is incorrect. (Exception from HRESULT: 0x80070057 (E_INVALIDARG))```
3. After dismiss with Ok, the project is created (without VSIX, only analyzer + test project)
4. From the package manager console: `Update-Package microsoft.codeanalysis.csharp -pre`
```Update-Package : Could not install package 'Microsoft.CodeAnalysis.Common 2.0.0-rc2'. You are trying to install this package into a project that targets '.NETPortable,Version=v4.5,Profile=Profile7', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.```
5. Project properties, Library, Targeting, select Target .NET Platform Standard
```The project's targets cannot be changed. The selected targets require the project to opt-into NuGet 3.0 support, however, Visual Studio cannot automatically do this for you. Please uninstall all NuGet packages and try again.```

**Expected Behavior**:
Project is created and packages can be updated without errors.

**Actual Behavior**:
Various errors.

**Comment by Pilchie** [comment]
@jmarolf you're working on updating the templates, right?

**Comment by jmarolf** [comment]
Yes, I've just checked in some changes to fix this.  RC changes how the VSSDK works so we need to react to that.  @bkoelman once we have a signed build with the new VSSDK I'll respond here with a link

**Comment by bkoelman** [comment]
@jmarolf Any news on this? I'd like to help by trying out the new templates. This would give us the opportunity to identify issues before RTM.

**Comment by jmarolf** [comment]
@bkoelman the template fixes should be live now: https://marketplace.visualstudio.com/items?itemName=VisualStudioProductTeam.NETCompilerPlatformSDK if you have the time I would love for you to take a look

**Comment by bkoelman** [comment]
@jmarolf After installing this in VS2017 Community RC4, I'm not seeing the project templates. I have installed the Extensibility, Web and Windows workloads. Am I missing something?

![image](https://cloud.githubusercontent.com/assets/10324372/23085276/10025b8e-f567-11e6-89f9-a291f7076c96.png)

When trying to install via project template named "Download the .NET Compiler Platform SDK", it tells me that the SDK is already installed.

**Comment by bkoelman** [comment]
After installing this in VS2015.3 Community Edition, the templates are gone too.

**Comment by jmarolf** [comment]
@bkoelman change your .NET Framework version to be 4.6 instead of 4.5.2

**Comment by bkoelman** [comment]
Thanks, that helps. Perhaps a note at https://marketplace.visualstudio.com/items?itemName=VisualStudioProductTeam.NETCompilerPlatformSDK about switching framework version would be helpful, as the default framework after VS installation is set to v4.5.2. Also the text there does not mention VS 2017.

Please see my findings below.

* The analyzer project file contains `<TargetFrameworkVersion>v4.5</TargetFrameworkVersion>` while others refer to v4.6. Is this exception intentional?

* The roslyn assembly references in the Test project have `<Private>True</Private>` in VS2015, while those lines are omitted in VS2017. In contrast, the analyzer project contains `<Private>False</Private>` for both VS versions.

* The Test project in VS2015 contains `<Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />`, which seems an optimization for loading the Test Explorer package. In VS2017, this line is missing.

* The VSIX project in VS2017 contains extra lines, presumably added by project upgrade. I always delete those, but I am not sure if that would have any side effects here.
```
    <FileUpgradeFlags>
    </FileUpgradeFlags>
    <OldToolsVersion>14.0</OldToolsVersion>
    <UpgradeBackupLocation />
    <PublishUrl>publish\</PublishUrl>
    <Install>true</Install>
    <InstallFrom>Disk</InstallFrom>
    <UpdateEnabled>false</UpdateEnabled>
    <UpdateMode>Foreground</UpdateMode>
    <UpdateInterval>7</UpdateInterval>
    <UpdateIntervalUnits>Days</UpdateIntervalUnits>
    <UpdatePeriodically>false</UpdatePeriodically>
    <UpdateRequired>false</UpdateRequired>
    <MapFileExtensions>true</MapFileExtensions>
    <ApplicationRevision>0</ApplicationRevision>
    <ApplicationVersion>1.0.0.%2a</ApplicationVersion>
    <IsWebBootstrapper>false</IsWebBootstrapper>
    <UseApplicationTrust>false</UseApplicationTrust>
    <BootstrapperEnabled>true</BootstrapperEnabled>
```
and
```
  <ItemGroup>
    <BootstrapperPackage Include=".NETFramework,Version=v4.6">
      <Visible>False</Visible>
      <ProductName>Microsoft .NET Framework 4.6 %28x86 and x64%29</ProductName>
      <Install>true</Install>
    </BootstrapperPackage>
    <BootstrapperPackage Include="Microsoft.Net.Framework.3.5.SP1">
      <Visible>False</Visible>
      <ProductName>.NET Framework 3.5 SP1</ProductName>
      <Install>false</Install>
    </BootstrapperPackage>
  </ItemGroup>
```

Furthermore, in VS2017 a .vsix.csproj.user file gets added with contents:
```
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup>
    <PublishUrlHistory />
    <InstallUrlHistory />
    <SupportUrlHistory />
    <UpdateUrlHistory />
    <BootstrapperUrlHistory />
    <ErrorReportUrlHistory />
    <FallbackCulture>en-US</FallbackCulture>
    <VerifyUploadedFiles>false</VerifyUploadedFiles>
  </PropertyGroup>
</Project>
```

All this can probably be prevented by setting `<MinimumVisualStudioVersion>15.0</MinimumVisualStudioVersion>` initially so that project upgrade will not kick in. I have not tried that.

* The updated SDK adds the next line at the top of all three project files:
```
<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
```
I think that should not be added, unless there is a requirement that each analyzer must be licensed under Apache License. Either way, the project template does not contain a file License.txt that the comment refers to.

**Comment by bkoelman** [comment]
Additionally, the Test project does not reference the Microsoft.CodeAnalysis.Analyzers package. But when doing anything with the packages in the test project (for example: `update-package -reinstall Microsoft.CodeAnalysis`), it automatically adds 1.0 because there seems to be a dependency to it. After the NuGet operation, the user experiences the "ADxxx Analyzer threw exception" error, for which we created 1.1.0 to resolve that.

So to summarize, I would recommend adding a package reference to Microsoft.CodeAnalysis.Analyzers v1.1 in the Test project.

**Comment by bkoelman** [comment]
The .vsixmanifest contains a line referring to .NET framework 4.5:
```
<Dependency Id="Microsoft.Framework.NDP" DisplayName="Microsoft .NET Framework" d:Source="Manual" Version="[4.5,)" />
```

Consider changing this also to 4.6, if that makes sense.

**Comment by bkoelman** [comment]
Something else: when creating a new analyzer project, then trying to update packages to latest rc4 bits, updating the Test project succeeds, but the analyzer project fails with:
```
Could not install package 'Microsoft.CodeAnalysis.Common 2.0.0-rc4'. You are trying to install this package into a project that targets '.NETPortable,Version=v4.5,Profile=Profile7', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.
```

**Comment by Pilchie** [comment]
That last one is because the 2.0 packages only target .Net Standard. You'll need an "imports" statement, but I'm not sure how to do that in packages.config of the top of my head. 

Soon we will likely release a .Net Standard based template for analyzers for C# that should help with some of these.

**Comment by jmarolf** [comment]
Thanks @bkoelman I think these issues need to be fixed:

- The updated SDK adds the copyright line at the top of all three project files
- I would recommend adding a package reference to Microsoft.CodeAnalysis.Analyzers v1.1 in the Test project.
- The .vsixmanifest contains a line referring to .NET framework 4.5
- When creating a new analyzer project, then trying to update packages to latest rc4 bits, updating the Test project succeeds, but the analyzer project fails

regarding your other questions:

*The analyzer project file contains <TargetFrameworkVersion>v4.5</TargetFrameworkVersion> while others refer to v4.6. Is this exception intentional?*
The analyzer project is Portable7 which is 4.5 compatible.  The workspace APIs referenced in the test project only run on 4.6

*The Test project in VS2015 contains <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />, which seems an optimization for loading the Test Explorer package. In VS2017, this line is missing.*
This reference is automatically added by the test discovery service.  I decided to remove this for terseness.

*All this can probably be prevented by setting <MinimumVisualStudioVersion>15.0</MinimumVisualStudioVersion> initially so that project upgrade will not kick in. I have not tried that.*
I want to keep it at 14.0 as these templates are still used in Visual Studio 2015

**Comment by bkoelman** [comment]
@jmarolf Thanks for explaining, sounds like a good plan!

Just to make sure: the CopyLocal settings is not something that needs attention?

**Comment by jmarolf** [comment]
@bkoelman I'll verify that only the user dlls are deployed to the VSIX.  CopyLocal should be set so that the user does not deploy roslyn dependencies with their VSIX.

**Comment by mqueirozcorreia** [comment]
+1 here, waiting for the solution
